File: src/lib/permissions.ts
import { getStoredUser, type AuthUser } from '@/lib/auth';
import { apiFetch } from '@/lib/api-client';

export type ModulePerm = { view: boolean; write: boolean };
export type PermissionsPayload = {
  role: string;
  modules: Record<string, ModulePerm>;
};

const CACHE_KEY = 'geohazard_permissions';

/** 路径前缀 → 权限模块 */
const PATH_MODULE: { prefix: string; module: string }[] = [
  { prefix: '/system/users', module: 'users' },
  { prefix: '/system/config', module: 'system' },
  { prefix: '/hazard', module: 'hazard' },
  { prefix: '/monitoring', module: 'monitoring' },
  { prefix: '/warning', module: 'warning' },
  { prefix: '/emergency', module: 'emergency' },
  { prefix: '/ecology', module: 'ecology' },
  { prefix: '/equipment', module: 'equipment' },
  { prefix: '/geology', module: 'geology' },
  { prefix: '/inspection', module: 'inspection' },
  { prefix: '/statistics', module: 'statistics' },
  { prefix: '/dashboard', module: 'dashboard' },
  { prefix: '/screen', module: 'dashboard' },
];

export function moduleForPath(path: string): string | null {
  for (const row of PATH_MODULE) {
    if (path === row.prefix || path.startsWith(row.prefix + '/')) return row.module;
  }
  return null;
}

export async function fetchMyPermissions(): Promise<PermissionsPayload> {
  const data = await apiFetch<PermissionsPayload>('/api/users/permissions/');
  if (typeof window !== 'undefined') {
    localStorage.setItem(CACHE_KEY, JSON.stringify(data));
  }
  return data;
}

export function getCachedPermissions(): PermissionsPayload | null {
  if (typeof window === 'undefined') return null;
  const raw = localStorage.getItem(CACHE_KEY);
  if (!raw) return null;
  try {
    return JSON.parse(raw) as PermissionsPayload;
  } catch {
    return null;
  }
}

export function canViewModule(module: string, perms?: PermissionsPayload | null, user?: AuthUser | null): boolean {
  const u = user ?? getStoredUser();
  if (u?.role === 'admin') return true;
  const p = perms ?? getCachedPermissions();
  if (p?.modules?.[module]) return Boolean(p.modules[module].view);
  // 无缓存时按角色保守放行只读业务页
  if (module === 'users') return u?.role === 'admin';
  if (module === 'system') return u?.role === 'admin' || u?.role === 'leader';
  return true;
}

export function canWriteModule(module: string, perms?: PermissionsPayload | null, user?: AuthUser | null): boolean {
  const u = user ?? getStoredUser();
  if (u?.role === 'admin') return true;
  const p = perms ?? getCachedPermissions();
  if (p?.modules?.[module]) return Boolean(p.modules[module].write);
  if (u?.role === 'viewer') return false;
  if (u?.role === 'grid_worker') return module === 'inspection' || module === 'hazard';
  return u?.role === 'leader' || u?.role === 'operator';
}

export function canViewPath(path: string): boolean {
  const mod = moduleForPath(path);
  if (!mod) return true;
  return canViewModule(mod);
}

File: backend/users/permissions.py
"""基于角色的细粒度权限"""
from __future__ import annotations

from rest_framework.permissions import BasePermission, SAFE_METHODS

# 模块 → 允许写操作的角色
WRITE_ROLES = {
    'users': {'admin'},
    'hazard': {'admin', 'leader', 'operator'},
    'monitoring': {'admin', 'leader', 'operator'},
    'warning': {'admin', 'leader', 'operator'},
    'emergency': {'admin', 'leader', 'operator'},
    'ecology': {'admin', 'leader', 'operator'},
    'equipment': {'admin', 'leader', 'operator'},
    'geology': {'admin', 'leader', 'operator'},
    'platform': {'admin', 'leader'},
    'inspection': {'admin', 'leader', 'operator', 'grid_worker'},
    'dashboard': {'admin', 'leader', 'operator', 'grid_worker', 'viewer'},
    'statistics': {'admin', 'leader', 'operator', 'viewer'},
}

# 菜单/模块可见角色
VIEW_ROLES = {
    'users': {'admin'},
    'hazard': {'admin', 'leader', 'operator', 'grid_worker', 'viewer'},
    'monitoring': {'admin', 'leader', 'operator', 'viewer'},
    'warning': {'admin', 'leader', 'operator', 'viewer'},
    'emergency': {'admin', 'leader', 'operator', 'viewer'},
    'ecology': {'admin', 'leader', 'operator', 'viewer'},
    'equipment': {'admin', 'leader', 'operator', 'viewer'},
    'geology': {'admin', 'leader', 'operator', 'viewer'},
    'platform': {'admin', 'leader', 'operator', 'viewer'},
    'inspection': {'admin', 'leader', 'operator', 'grid_worker', 'viewer'},
    'dashboard': {'admin', 'leader', 'operator', 'grid_worker', 'viewer'},
    'statistics': {'admin', 'leader', 'operator', 'viewer'},
    'system': {'admin', 'leader'},
}


def resolve_module(path: str) -> str | None:
    p = path or ''
    mapping = (
        ('/api/users', 'users'),
        ('/api/hazard', 'hazard'),
        ('/api/monitoring', 'monitoring'),
        ('/api/warning', 'warning'),
        ('/api/emergency', 'emergency'),
        ('/api/ecology', 'ecology'),
        ('/api/equipment', 'equipment'),
        ('/api/geology', 'geology'),
        ('/api/platform', 'platform'),
        ('/api/dashboard', 'dashboard'),
        ('/api/statistics', 'statistics'),
    )
    for prefix, mod in mapping:
        if p.startswith(prefix):
            return mod
    return None


def role_can_view(role: str, module: str) -> bool:
    if role == 'admin':
        return True
    return role in VIEW_ROLES.get(module, set())


def role_can_write(role: str, module: str) -> bool:
    if role == 'admin':
        return True
    return role in WRITE_ROLES.get(module, set())


def permissions_payload(role: str) -> dict:
    modules = sorted(set(VIEW_ROLES) | set(WRITE_ROLES))
    return {
        'role': role,
        'modules': {
            m: {
                'view': role_can_view(role, m),
                'write': role_can_write(role, m),
            }
            for m in modules
        },
    }


class RoleActionPermission(BasePermission):
    """
    已登录可读（按模块可见性）；写操作校验 WRITE_ROLES。
    ingest 等 AllowAny 视图不受影响。
    """

    message = '当前角色无权执行此操作'

    def has_permission(self, request, view):
        user = request.user
        if not user or not user.is_authenticated:
            return False
        role = getattr(user, 'role', '') or 'viewer'
        if role == 'admin' or user.is_superuser:
            return True

        module = resolve_module(request.path)
        if not module:
            return True  # 未知模块保持登录即可

        if request.method in SAFE_METHODS:
            return role_can_view(role, module)

        # 部分只读动作
        action = getattr(view, 'action', None)
        if action in (
            'export', 'statistics', 'next_code', 'latest', 'related',
            'map_geojson', 'shelters', 'permissions',
        ):
            return role_can_view(role, module)

        return role_can_write(role, module)
